src/libcollections/btree/map.rs

```diff
@@ -1630,32 +1806,61 @@ fn handle_underfull_node<'a, K, V>(node: NodeRef<marker::Mut<'a>,
     };
 
     if handle.can_merge() {
-        return Merged(handle.merge().into_node());
+        Merged(handle.merge().into_node())
     } else {
-        unsafe {
-            let (k, v, edge) = if is_left {
-                handle.reborrow_mut().left_edge().descend().pop()
-            } else {
-                handle.reborrow_mut().right_edge().descend().pop_front()
-            };
-
-            let k = mem::replace(handle.reborrow_mut().into_kv_mut().0, k);
-            let v = mem::replace(handle.reborrow_mut().into_kv_mut().1, v);
-
-            // FIXME: reuse cur_node?
-            if is_left {
-                match handle.reborrow_mut().right_edge().descend().force() {
-                    Leaf(mut leaf) => leaf.push_front(k, v),
-                    Internal(mut internal) => internal.push_front(k, v, edge.unwrap())
-                }
-            } else {
-                match handle.reborrow_mut().left_edge().descend().force() {
-                    Leaf(mut leaf) => leaf.push(k, v),
-                    Internal(mut internal) => internal.push(k, v, edge.unwrap())
-                }
-            }
+        if is_left {
+            Stole(handle.steal_left().into_node())
+        } else {
+            Stole(handle.steal_right().into_node())
         }
+    }
+}
 
-        return Stole(handle.into_node());
+// Helper enum for MergeIter
+enum MergeResult {
+    Left,
+    Right,
+    Both,
+    None,
+}
+
+impl<K: Ord, V, I: Iterator<Item=(K, V)>> Iterator for MergeIter<K, V, I> {
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<(K, V)> {
+        let res = match (&self.left_cur, &self.right_cur) {
+            (&Some((ref left_key, _)), &Some((ref right_key, _))) => {
+                match left_key.cmp(right_key) {
+                    Ordering::Less => MergeResult::Left,
+                    Ordering::Equal => MergeResult::Both,
+                    Ordering::Greater => MergeResult::Right,
+                }
+            },
+            (&Some(_), &None) => MergeResult::Left,
+            (&None, &Some(_)) => MergeResult::Right,
+            (&None, &None) => MergeResult::None,
+        };
```

Comment:
I think you can get rid of MergeResult with an early return, and use Ordering instead.
```rust
(&None, &None) => {
    return None
}
```